File: games/marble_run/native-resources/android-src/app/src/main/java/com/basegamelab/marblerun/sdk/AppsFlyerAttributionPlugin.java
package com.basegamelab.marblerun.sdk;

import com.appsflyer.AppsFlyerLib;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;
import org.json.JSONObject;

@CapacitorPlugin(name = "AppsFlyerAttribution")
public class AppsFlyerAttributionPlugin extends Plugin {

    private boolean initialized = false;

    @PluginMethod
    public void initialize(PluginCall call) {
        String devKey = trimmed(call.getString("devKey"));
        if (devKey == null) {
            resolveBoolean(call, "initialized", false);
            return;
        }
        if (initialized) {
            resolveBoolean(call, "initialized", true);
            return;
        }
        AppsFlyerLib lib = AppsFlyerLib.getInstance();
        lib.setDebugLog(Boolean.TRUE.equals(call.getBoolean("debugLogging", false)));
        lib.init(devKey, null, getContext().getApplicationContext());
        lib.start(getContext().getApplicationContext());
        initialized = true;
        resolveBoolean(call, "initialized", true);
    }

    @PluginMethod
    public void trackEvent(PluginCall call) {
        String eventName = trimmed(call.getString("eventName"));
        if (!initialized || eventName == null) {
            resolveBoolean(call, "tracked", false);
            return;
        }
        Map<String, Object> values = new HashMap<>();
        JSONObject raw = call.getObject("eventValues", new JSObject());
        Iterator<String> keys = raw.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            Object value = raw.opt(key);
            if (value instanceof String) {
                values.put(key, capped((String) value));
            }
        }
        AppsFlyerLib.getInstance().logEvent(getContext().getApplicationContext(), eventName, values);
        resolveBoolean(call, "tracked", true);
    }

    @PluginMethod
    public void getStatus(PluginCall call) {
        JSObject status = new JSObject();
        status.put("initialized", initialized);
        if (initialized) {
            status.put("appsFlyerId", AppsFlyerLib.getInstance().getAppsFlyerUID(getContext().getApplicationContext()));
        } else {
            status.put("appsFlyerId", JSObject.NULL);
        }
        call.resolve(status);
    }

    private static void resolveBoolean(PluginCall call, String key, boolean value) {
        JSObject result = new JSObject();
        result.put(key, value);
        call.resolve(result);
    }

    private static String trimmed(String value) {
        if (value == null) return null;
        String out = value.trim();
        return out.isEmpty() ? null : out;
    }

    private static String capped(String value) {
        return value.length() <= 96 ? value : value.substring(0, 96);
    }
}

File: games/find_the_bird/native-resources/ios/App/AppsFlyerAttributionPlugin.swift
import Capacitor
import Foundation
import UIKit
import AppsFlyerLib

@objc(AppsFlyerAttributionPlugin)
public final class AppsFlyerAttributionPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "AppsFlyerAttributionPlugin"
    public let jsName = "AppsFlyerAttribution"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "trackEvent", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getStatus", returnType: CAPPluginReturnPromise),
    ]
    private var initialized = false
    private var activeObserver: NSObjectProtocol?

    deinit {
        if let activeObserver {
            NotificationCenter.default.removeObserver(activeObserver)
        }
    }

    @objc func initialize(_ call: CAPPluginCall) {
        // Serialize initialization with UIKit lifecycle notifications.
        DispatchQueue.main.async { self.initializeOnMain(call) }
    }

    private func initializeOnMain(_ call: CAPPluginCall) {
        guard !initialized else { call.resolve(["initialized": true]); return }
        guard let devKey = call.getString("devKey"), !devKey.isEmpty,
              let appleAppId = call.getString("appleAppId"), !appleAppId.isEmpty else {
            call.resolve(["initialized": false]); return
        }
        let blockedPartners = call.getArray("blockedPartners", String.self) ?? []
        let sdk = AppsFlyerLib.shared()
        sdk.appsFlyerDevKey = devKey
        sdk.appleAppID = appleAppId
        sdk.isDebug = call.getBool("debugLogging") ?? false
        // General-audience policy: partners activated in the AppsFlyer dashboard
        // receive install and event postbacks. Only explicitly blocked partners
        // are filtered, and the filter is applied before start.
        if !blockedPartners.isEmpty {
            sdk.setSharingFilterForPartners(blockedPartners)
        }
        initialized = true
        activeObserver = NotificationCenter.default.addObserver(
            forName: UIApplication.didBecomeActiveNotification,
            object: nil,
            queue: .main
        ) { [weak self] _ in
            guard self?.initialized == true else { return }
            AppsFlyerLib.shared().start()
        }
        // The bridge can initialize after the first active notification. If it
        // initializes while inactive, the observer supplies the first start.
        if UIApplication.shared.applicationState == .active {
            sdk.start()
        }
        call.resolve(["initialized": true])
    }

    @objc func trackEvent(_ call: CAPPluginCall) {
        guard initialized, let name = call.getString("eventName") else {
            call.resolve(["tracked": false]); return
        }
        let values = call.getObject("eventValues") ?? [:]
        AppsFlyerLib.shared().logEvent(name: name, values: values) { _, _ in }
        call.resolve(["tracked": true])
    }

    @objc func getStatus(_ call: CAPPluginCall) {
        call.resolve(["initialized": initialized, "appsFlyerId": initialized ? AppsFlyerLib.shared().getAppsFlyerUID() : NSNull()])
    }
}
